File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        curr = TreeNode()
        stack = deque([root]) if root else None
        while stack:
            temp = stack.pop()
            if temp.right: stack.append(temp.right)
            if temp.left: stack.append(temp.left)
            curr.right = temp
            curr.left = None
            curr = curr.right            
        

File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode *root) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);
        
        while (root) {
            if (root->left && root->right) {
                TreeNode* t = root->left;
                while (t->right) t = t->right;
                t->right = root->right;
            }

            if (root->left) root->right = root->left;
            root->left = nullptr;
            root = root->right;
        }
    }
};